libs/platform/ebpf_object.c

```diff
@@ -326,3 +329,35 @@ ebpf_object_dereference_by_id(ebpf_id_t id, ebpf_object_type_t object_type)
     ebpf_lock_unlock(&_ebpf_object_tracking_list_lock, state);
     return return_value;
 }
+
+ebpf_result_t
+ebpf_object_reference_by_handle(
+    ebpf_handle_t handle, ebpf_object_type_t object_type, _Outptr_ ebpf_core_object_t** object)
+{
+    ebpf_result_t result;
+    ebpf_core_object_t* local_object = NULL;
+    *object = NULL;
+
+    result = ebpf_reference_object_by_handle(handle, (ebpf_base_object_t**)&local_object);
+    if (result == EBPF_SUCCESS) {
+        __analysis_assume(local_object != NULL);
+        if (local_object->base.marker != _ebpf_object_marker) {
+            result = EBPF_INVALID_OBJECT;
+            goto Exit;
+        }
+
+        if ((object_type != EBPF_OBJECT_UNKNOWN) && (ebpf_object_get_type(local_object) != object_type)) {
+            result = EBPF_INVALID_OBJECT;
+            goto Exit;
+        }
```

Comment:
By moving this check outside ebpf_reference_object_by_handle, it entails an extra interlocked increment and decrement.
What are we buying by doing it outside instead of inside?

Reply:
`ebpf_reference_object_by_handle()` is now generic as it is used for both `ebpf_core_object_t` and `ebpf_native_module_t` objects
An "object_type" is only applicable to objects derived from `ebpf_core_object_t` (`ebpf_core_map`, `ebpf_core_program`, etc.)

Hence I had to move this check out of `ebpf_reference_object_by_handle`.

Reply:
Can this happen in a hot path?  (If so, I'd prefer leaving a variation that doesn't require extra interlocked operations)

Reply:
Changed the code to first validate the type and then acquire reference on the object.